### DISCUSSION TITLE
C vs C++ interface

### DISCUSSION THREAD

**Discussion by charleskhazoom** [casadi/casadi#3309] [Q&A] [answered]
I have a big casadi function constructed in Matlab. I code generate it and load back the .so file using casadi.external(). Then, what is the recommended way to run it in c or c++ as fast as possible ? I see two options : 

1. use the c API of casadi as in the [c_api_usage example](https://github.com/casadi/casadi/blob/main/docs/examples/cplusplus/c_api_usage.cpp) 
2. load the function in c++ and call the function using casadi c++ using casadi::Function::load()

are those two options equivalent in terms of speed ?

**Comment by jgillis** [comment]
Classic api of codegen is most appropriate:
docs/examples/cplusplus/codegen_usage.cpp

**Reply by charleskhazoom** [reply]
Thank you. So the example [codegen_usage.cpp](https://github.com/casadi/casadi/blob/main/docs/examples/cplusplus/codegen_usage.cpp), shows 3 possible ways :

1. usage_c();
2. usage_cplusplus();
3. usage_c_with_mem();

Is there one that's better, or are they all equally appropriate ?

**Reply by jgillis** [reply, marked as answer]
For speed, you want `usage_c` or `usage_c_with_mem`. `usage_c` gives more control
